File: 13a.py
#! /usr/bin/python3
import sys
import re

def _parse(tokens):
    struct = []
    while (tokens):
        if re.match(r'[0-9]', tokens[0]):
            struct += [int(tokens.pop(0))]

        elif '[' == tokens[0]:
            parsed, rest = _parse(tokens[1:])
            struct += [parsed]
            tokens = rest

        elif ']' == tokens[0]:
            return struct, tokens[1:]

        else:
            raise Exception(f"Can't parse {tokens[0]}")
    return struct[0]

def parse(packet):
    tokens = re.findall(r'([0-9]+|[\[\]])', line)
    return _parse(tokens)

def compare(packets):
    while not packets[1] is None and len(packets[1]):
        if not len(packets[0]):
            return 1

        p1 = packets[0].pop(0)
        p2 = packets[1].pop(0)

        if not type(p1) is list and not type(p2) is list:
            if p1 != p2:
                return 1 if p1 < p2 else 0

        elif not type(p1) is list or not type(p2) is list:
            packets[1].insert(0, p2 if type(p2) is list else [p2])
            packets[0].insert(0, p1 if type(p1) is list else [p1])

        else:
            comp = compare([p1, p2])
            if not comp is None:
                return comp
    if len(packets[0]):
        return 0

    return None

i = 0
total = 0
packets = [False, False]
with open(sys.argv[1], 'r') as fh:
    for line in fh:
        if 1 == len(line):
            continue

        packets[1 if not packets[0] is False else 0] = parse(line)
        if not packets[1] is False:
            i += 1
            if compare(packets) == 1:
                total += i
            packets = [False, False]

print(total)

File: 13a.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };
use experimental 'signatures';

use ARGV::OrDATA;

sub parse($packet) {
    my @list;
    while ($packet =~ s/((?:,?[0-9]+)+|[\[\]])//) {
        my $token = $1;
        if ($token =~ /[0-9]/) {
            $token =~ s/^,+//;
            push @list, split /,/, $token;

        } elsif ('[' eq $token) {
            my ($parsed, $rest) = parse($packet);
            push @list, $parsed;
            $packet = $rest;

        } elsif (']' eq $token) {
            return \@list, $packet

        } else {
            die "Can't parse '$packet'";
        }
    }
    return @list
}

sub compare(@packets) {
    while (@{ $packets[1] }) {
        my $p1 = shift @{ $packets[0] };
        my $p2 = shift @{ $packets[1] };
        return 1 unless defined $p1;

        if (! ref $p1 && ! ref $p2) {
            return $p1 < $p2 ? 1 : 0 if $p1 != $p2;

        } elsif (! ref $p1 || ! ref $p2) {
            unshift @{ $packets[1] }, ref $p2 ? $p2 : [$p2];
            unshift @{ $packets[0] }, ref $p1 ? $p1 : [$p1];
        } else {
            my $cmp = compare($p1, $p2);
            return $cmp if defined $cmp;
        }
    }
    return 0 if @{ $packets[0] };

    return
}

my $sum = 0;
local $/ = "";
while (<>) {
    my @packets = map { parse($_) } split;
    if (compare(@packets)) {
        $sum += $.;
    }
}

say $sum;

__DATA__
[1,1,3,1,1]
[1,1,5,1,1]

[[1],[2,3,4]]
[[1],4]

[9]
[[8,7,6]]

[[4,4],4,4]
[[4,4],4,4,4]

[7,7,7,7]
[7,7,7]

[]
[3]

[[[]]]
[[]]

[1,[2,[3,[4,[5,6,7]]]],8,9]
[1,[2,[3,[4,[5,6,0]]]],8,9]
